whelktool/scripts/cleanups/2021/04/lxl-3547-link-expressionOf.groovy

```diff
@@ -0,0 +1,194 @@
+ * expressionOf with language can link to work with same language or no language
+ * expressionOf without language can link to work without language
+ * 
+ * See LXL-3547 for more info
+ * 
+ * TODO: decide which non-linkable expressionOf should be extracted to new works and linked, i.e. how many identical expressionOf have to exist 
+ */
+import org.apache.commons.lang3.StringUtils
+import whelk.util.Unicode
+import java.text.Normalizer
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.regex.Pattern
+
+PrintWriter linked = getReportWriter("linked.txt")
+PrintWriter noLang = getReportWriter("no-lang.txt")
+PrintWriter multiMatch = getReportWriter("multiple-matches.txt")
+PrintWriter sameExpr = getReportWriter("same-expr.txt")
+
+uniformWorks = getUniformWorks()
+
+def q = [
+        '@type'                                            : ['Instance'],
+        'exists-instanceOf.expressionOf.hasTitle.mainTitle': ['true'],
+        'exists-instanceOf.expressionOf.@id'               : ['false'],
+        '_sort'                                            : ['@id'],
+]
+
+def notLinkedExpr = new ConcurrentHashMap<Map, ConcurrentLinkedQueue<Map>>()
+selectByIds(queryIds(q).collect()) { bib -> 
+    List<Map> expressionOf = asList(getPathSafe(bib.doc.data, ['@graph', 1, 'instanceOf', 'expressionOf']))
+
+    if (!expressionOf) {
+        return
+    }
+    
+    boolean modified = false
+    expressionOf.each { Map e -> 
+        if (e['@id']) {
+            return
+        }
+
+        // there is always exactly one title
+        def cmpMap = Norm.cmpTitle(asList(e['hasTitle']).first()) + Norm.cmpOther(e)
+        def works = uniformWorks.getOrDefault(cmpMap, Collections.emptyList())
+        
+        incrementStats('number of works found for expressionOf', works.size())
+        
+        if (works) {
+            def worksLang = compatibleLanguages(e, works)
+            incrementStats('number of works found for expressionOf, filtered on lang', worksLang.size())
+
+            if (worksLang.size() == 0) {
+                noLang.println("${bib.doc.id} ${toString(e)} ??? ${works.collect {it['@id'] + " " + toString(it)}}")
+            }
+            else if (worksLang.size() == 1) {
+                linked.println("${bib.doc.id} ${toString(e)} --> ${worksLang[0].with {it['@id'] + " " + toString(it)}}")
+                e.clear()
+                e['@id'] = worksLang[0]['@id']
+                modified |= true
+            }
+            else {
+                multiMatch.println("${bib.doc.id} ${toString(e)} ??? ${worksLang.collect {it['@id'] + " " + toString(it)}}")
+            }
+        }
+        else {
+            def contribution = asList(getPathSafe(bib.doc.data, ['@graph', 1, 'instanceOf', 'expressionOf'])).findAll{ it['@type'] == 'PrimaryContribution'}
+            def ee = contribution ? e + ['contribution': contribution] : e
```

Comment:
What is the purpose of this part? Not sure that I understand.

Reply:
Keeping track of expressionOf that were not linked. Including contribution from the "main" work. Identical ones with many copies could be extracted to new records.

Reply:
I see! But shouldn't the path to the "main" work contribution be just ['@graph', 1, 'instanceOf']?

Reply:
Forget that, the "main" work is of course the entity in 'expressionOf', and different expressions of it can have different contributions while still pointing to the same main work right?